design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -541,28 +564,51 @@ The following naming strategies were rejected:
+projection _without_ their source qualifier, for example:
 
-`JOINKEY` or some other hardcoded name was rejected primarliy due to the implementation issues
-mentioned above: not only would it be possible to validate the presence of `JOINKEY` in the
-projection, the user would also be free to use `JOINKEY` in other places, e.g. as a parameter to a
-UDF or within a binary expression, etc.  This would not work by default and would require extra
-effort to make work, including additional steps in the streams topology - there seems little ROI on
-such effort or the complexity it will add to the code.
+```sql
+-- I1 schema: ID KEY, V0, V1
+-- I2 schema: ID KEY, V0, V2
+CREATE TABLE OUTPUT AS
+   SELECT I1.ID, I2.V0, V1, V2 FROM I1 JOIN I2 ON I1.ID = I2.ID;
+-- OUTPUT schema: I1_ID KEY, I2_V0, V1, V2
+```
+
+In the above query, note how `I1.ID` and `I2.V0` must include their source to avoid being ambiguous,
+however `V1` and `V2` do not.
 
-#### `KSQL_COL_x`
+Now let us consider how this would play out if synthetic join keys had a hardcoded name `ROWKEY`:
 
-The synthesised column would take on a generated name in the form `KSQL_COL_x`, where `x` is an
-integer and the name is guaranteed to be unique, i.e. to not clash with any other system generated
-column names.
+```sql
+CREATE TABLE TABLE_WITH_ROWKEY AS
+    SELECT * FROM I1 FULL OUTER JOIN I2 ON I1.ID = I2.ID;
+-- T1: ROWKEY KEY, I1_ID, I1_V0, I1_V1, I2_ID, I2_V0, I2_V1
+
+CREATE TABLE OUTPUT AS
+    SELECT ROWKEY, [...] FROM TABLE_WITH_ROWKEY I3 FULL OUTER JOIN I4 ON I3.ROWKEY = I4.ID;
+```
 
-This was rejected as the name would be hard for the user to know by looking at the query, i.e. the
-presence of other columns with generated names may affect the name of the synthesised column. This
-was deemed to make this solution unworkable.
+The first statement executes without issue and results in a table with a schema that includes the
+hardcoded `ROWKEY` column name. However, the presence of `ROWKEY` second statement's projection
+posses a problem: does this `ROWKEY` pertain to the synthetic join key this query introduces or
+`TABLE_WITH_ROWKEY.ROWKEY`?
+
+One could argue that a `ROWKEY` without a source would always refer to the newly synthesised join
+column, and the user is free to include a qualified `TABLE_WITH_ROWKEY.ROWKEY` if they require the
+source column.  However, this is currently rejected for the following reasons:
+
+  1. this breaks from the normal rules used to handle ambiguous column names, adding confusion
+     for users.
+  2. intuitively, users looking at the second statement would assume the `ROWKEY` in the projection
+     came from the `TABLE_WITH_ROWKEY`, i.e. its not clear its refering to a synthesised join column.
+  3. the current ambiguous column detection code isn't up to the job of handling this join case and
+     enhancement would require a large degree of redesign and movement of the logic within the
+     query analysis and planned stages.
```

Comment:
cc @derekjn @mjsax 

Chaps, I've brought the KLIP inline with the implementation in this PR.  

We also discussed using `ROWKEY` as a hardcoded name for the synthesised column name.  I've added some details to the KLIP explain the complexities and challenges of doing so.

Personally, I don't like the idea of using a hardcoded `ROWKEY` and introducing the ambiguity and inconsistencies I've hopefully explained above, naming:
  1.  It's hard for a user to know if `ROWKEY` in a projection refers to a synthesised column or a column from the source.
  2.  It's inconsistent with how ksqlDB has historically supported ambiguous column names in the projection. ksqlDB currently requires any ambiguous column name to be fully qualified, however the synthetic column can not be fully qualified as it does not come from a source table/stream.
  3. Implementing this will be a PITA to be honest, due to the current implementation we've inherited.

The ways forward I see are (in no particular order):
   a. Go with the hardcoded name in the upcoming release, with the likely release delays that will result..
   a. Stick with the system generated names in this PR, e.g. `KSQL_COL_0`.
   a. Go with the system generated names for the upcoming release and switch to hardcoded for the next.
   a. Require the user to _always_ provide an alias for any synthetic column, thereby avoiding the name clashing / ambiguity.  We've have to pick a new reserved column name, as `ROWKEY` is still the default key column name if one is not explicitly provided.  The main downside of such an approach is that a `SELECT *` would fail as it doesn't provide an alias.
   a. Add a unique postfix to  `ROWKEY` to ensure its unique. Most of the time it will be `ROWKEY`, but if the source data contains a `ROWKEY` column then it will become `ROWKEY_1`, or some other number to ensure its unique. This is essentially the same pattern as what this PR has, just using a different name, i.e. `ROWKEY` rather than `KSQL_COL`.  This would be easy to change to.

Your thoughts please gentlemen....

Reply:
Closing the loop on this ... choice... d was made